modin/core/dataframe/algebra/binary.py

```diff
@@ -48,32 +49,70 @@ def coerce_int_to_float64(dtype: np.dtype) -> np.dtype:
         return dtype
 
 
-def compute_dtypes_common_cast(first, second) -> np.dtype:
+def maybe_compute_dtypes_common_cast(
+    first, second, trigger_computations=False
+) -> Optional[pandas.Series]:
     """
     Precompute data types for binary operations by finding common type between operands.
 
     Parameters
     ----------
     first : PandasQueryCompiler
         First operand for which the binary operation would be performed later.
-    second : PandasQueryCompiler
+    second : PandasQueryCompiler, list-like or scalar
         Second operand for which the binary operation would be performed later.
+    trigger_computations : bool, default: False
+        Whether to trigger computation of the lazy metadata for `first` and `second`.
+        If False is specified this method will return None if any of the operands doesn't
+        have materialized dtypes.
 
     Returns
     -------
-    dtypes
-        The pandas series with precomputed dtypes.
+    pandas.Series
+        The pandas series with precomputed dtypes or None if there's not enough metadata to compute it.
 
     Notes
     -----
     The dtypes of the operands are supposed to be known.
     """
+    if not trigger_computations:
+        if not first._modin_frame.has_materialized_dtypes:
+            return None
+
+        if (
+            isinstance(second, type(first))
+            and not second._modin_frame.has_materialized_dtypes
+        ):
+            return None
+
     dtypes_first = first._modin_frame.dtypes.to_dict()
-    dtypes_second = second._modin_frame.dtypes.to_dict()
-    columns_first = set(first.columns)
-    columns_second = set(second.columns)
-    common_columns = columns_first.intersection(columns_second)
-    mismatch_columns = columns_first.union(columns_second) - common_columns
+    if isinstance(second, type(first)):
+        dtypes_second = second._modin_frame.dtypes.to_dict()
+        columns_first = set(first.columns)
+        columns_second = set(second.columns)
+        common_columns = columns_first.intersection(columns_second)
+        mismatch_columns = columns_first.union(columns_second) - common_columns
+    else:
+        if isinstance(second, (dict, list, tuple)):
+            second_dtypes_list = [
+                type(value)
+                for value in (second.values() if isinstance(second, dict) else second)
+            ]
+        elif is_scalar(second) or isinstance(second, np.ndarray):
+            second_dtypes_list = [getattr(second, "dtype", type(second))] * len(
+                dtypes_first
+            )
+        else:
+            raise NotImplementedError(
+                f"Can't compute common type for {type(first)} and {type(second)}."
+            )
+
+        dtypes_second = {
+            key: second_dtypes_list[idx] for idx, key in enumerate(dtypes_first.keys())
```

Comment:
we verify `other` lengths at front-end, so it shouldn't be the case
https://github.com/modin-project/modin/blob/baa860554febd58b7aa11947028c367664488887/modin/pandas/base.py#L294-L305

Reply:
Let's add this information to this function so that in the future we can better understand in what context it is worth using this function. (asserts + comments)